File: src/common/dates.py
"""Date parsing helpers shared across bank extractors."""

from __future__ import annotations

import re

from common.types import StatementMeta

MONTHS = ("JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL", "AUG", "SEP", "OCT", "NOV", "DEC")
FULL_MONTHS = (
    "JANUARY",
    "FEBRUARY",
    "MARCH",
    "APRIL",
    "MAY",
    "JUNE",
    "JULY",
    "AUGUST",
    "SEPTEMBER",
    "OCTOBER",
    "NOVEMBER",
    "DECEMBER",
)
MONTH_DAY_DATE_RE = re.compile(rf"^({'|'.join(MONTHS)})\s+(\d{{1,2}})$", re.IGNORECASE)


def month_to_number(month_name: str) -> int:
    upper = month_name.upper()
    if upper in MONTHS:
        return MONTHS.index(upper) + 1
    if upper in FULL_MONTHS:
        return FULL_MONTHS.index(upper) + 1
    raise ValueError(f"Unknown month: {month_name}")


def month_day_to_iso(month_name: str, day: str, year: int) -> str:
    month = month_to_number(month_name)
    return f"{year:04d}-{month:02d}-{int(day):02d}"


def infer_year(meta: StatementMeta) -> int:
    if meta.statement_date:
        return int(meta.statement_date[:4])
    if meta.statement_period_end:
        return int(meta.statement_period_end[:4])
    return 2024


def year_for_month_number(month_num: int, meta: StatementMeta) -> int:
    if meta.statement_period_start and meta.statement_period_end:
        start_y = int(meta.statement_period_start[:4])
        start_m = int(meta.statement_period_start[5:7])
        end_y = int(meta.statement_period_end[:4])
        end_m = int(meta.statement_period_end[5:7])
        if start_y == end_y:
            return start_y
        if month_num >= start_m:
            return start_y
        return end_y
    return infer_year(meta)


def normalize_month_day_date(raw: str, meta: StatementMeta) -> str:
    match = MONTH_DAY_DATE_RE.match(raw.strip())
    if not match:
        return raw.strip()
    month_num = month_to_number(match.group(1))
    day = match.group(2)
    year = year_for_month_number(month_num, meta)
    return f"{year:04d}-{month_num:02d}-{int(day):02d}"

File: common/dates.js
/** Shared date parsing helpers for bank statement extractors. */

export const MONTHS = ["JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL", "AUG", "SEP", "OCT", "NOV", "DEC"];
export const FULL_MONTHS = [
  "JANUARY",
  "FEBRUARY",
  "MARCH",
  "APRIL",
  "MAY",
  "JUNE",
  "JULY",
  "AUGUST",
  "SEPTEMBER",
  "OCTOBER",
  "NOVEMBER",
  "DECEMBER",
];
export const MONTH_DAY_DATE_RE = new RegExp(`^(${MONTHS.join("|")})\\s+(\\d{1,2})$`, "i");

const MONTH_TO_NUMBER = new Map([
  ...MONTHS.map((m, i) => [m, i + 1]),
  ...FULL_MONTHS.map((m, i) => [m, i + 1]),
]);

export function monthToNumber(monthName) {
  const num = MONTH_TO_NUMBER.get(monthName.toUpperCase());
  if (num === undefined) throw new Error(`Unknown month: ${monthName}`);
  return num;
}

export function monthDayToIso(monthName, day, year) {
  const month = monthToNumber(monthName);
  return `${String(year).padStart(4, "0")}-${String(month).padStart(2, "0")}-${String(parseInt(day, 10)).padStart(2, "0")}`;
}

export function inferYear(meta) {
  if (meta.statement_date) return parseInt(meta.statement_date.slice(0, 4), 10);
  if (meta.statement_period_end) return parseInt(meta.statement_period_end.slice(0, 4), 10);
  return 2024;
}

export function yearForMonthNumber(monthNum, meta) {
  if (meta.statement_period_start && meta.statement_period_end) {
    const startY = parseInt(meta.statement_period_start.slice(0, 4), 10);
    const startM = parseInt(meta.statement_period_start.slice(5, 7), 10);
    const endY = parseInt(meta.statement_period_end.slice(0, 4), 10);
    const endM = parseInt(meta.statement_period_end.slice(5, 7), 10);
    if (startY === endY) return startY;
    if (monthNum >= startM) return startY;
    return endY;
  }
  return inferYear(meta);
}

export function normalizeMonthDayDate(raw, meta) {
  const match = MONTH_DAY_DATE_RE.exec(raw.trim());
  if (!match) return raw.trim();
  const monthNum = monthToNumber(match[1]);
  const day = match[2];
  const year = yearForMonthNumber(monthNum, meta);
  return `${String(year).padStart(4, "0")}-${String(monthNum).padStart(2, "0")}-${String(parseInt(day, 10)).padStart(2, "0")}`;
}
